### PR TITLE
Feat: granular allowlist

### PR DESCRIPTION
**Pull request by TheDZhon**
## Summary

Implements granular `allowed_diffs` allowlists for both source and bytecode comparisons (closes #136).

Instead of blanket `--allow-source-diff`/`--allow-bytecode-diff` CLI flags that silently suppress all diffs for an address, diffs can now be declared precisely in config with mandatory `reason` fields and composable facets.

### Bytecode facets
- **`immutables`** — pin expected values at exact byte offsets
- **`cbor_metadata`** — allow Solidity CBOR metadata hash differences
- **`byte_ranges`** — allow arbitrary byte ranges
- **`constructor_args` / `constructor_calldata`** — override constructor parameters per-rule
- **`any`** — blanket allow (also suppresses RPC simulation errors)

### Source facets
- **`line_ranges`** — allow diffs in specific file regions (github/explorer line counts)
- **`files`** — allow diffs in specific files
- **`any`** — blanket allow

### Auto-suggestions
When diffs are detected but not covered by rules, the tool prints a ready-to-paste JSON `allowed_diffs` snippet with the exact facets needed. This makes adopting granular rules trivial — run once, copy the suggestion, fill in the `reason`.

## Key changes

| Area | What changed |
|---|---|
| **`diffyscan/utils/allowed_diffs.py`** | Validation, evaluation, normalization, and suggestion engine |
| **`diffyscan/diffyscan.py`** | Integrates allowlist evaluation into source/bytecode pipelines; prints suggestions in final summary; `"any"` bytecode rules now suppress RPC simulation errors |
| **`diffyscan/utils/common.py`** | Config loading validates `allowed_diffs` structure, boolean fields, and YAML hex quoting for allowlist addresses |
| **`diffyscan/utils/custom_types.py`** | `Config` TypedDict includes `allowed_diffs` and missing optional fields |
| **`pyproject.toml`** | Added mypy + type stubs to dev deps; `[tool.mypy]` config for incremental type checking |
| **Regression configs** (6 files) | Replaced all CLI `--allow-*-diff` flags with precise in-config `allowed_diffs` entries discovered via real runs against mainnet and hoodi RPCs |
| **`.github/workflows/regression.yml`** | Removed `flags` matrix, `DIFFYSCAN_FLAGS` env var, and bash flag-parsing; run command is now just `uv run diffyscan -Y -E -G "$DIFFYSCAN_CONFIG"` |

## Config example

```json
{
  "allowed_diffs": {
    "bytecode": {
      "0xd6A6...9426": [{
        "reason": "Escrow stores own deployed address as immutable",
        "immutables": [
          {"offset": 1093, "value": "0x000...9426"},
          {"offset": 6887, "value": "0x000...9426"}
        ]
      }]
    },
    "source": {
      "0x8aa3...9F3": [{
        "reason": "Import paths differ between GitHub and explorer-verified source",
        "line_ranges": [{
          "file": "contracts/CuratedSubmitExitRequestHashes.sol",
          "github": {"start": 6, "count": 7},
          "explorer": {"start": 6, "count": 7}
        }]
      }]
    }
  }
}
```

## Test plan

- [ ] 130 unit tests pass (`pytest -q`) — covers validation, evaluation, edge cases, suggestions, and config loading
- [ ] `mypy diffyscan/` — 0 errors across 16 source files
- [ ] All 12 regression matrix jobs pass (4 mainnet clean, 4 mainnet with allowed_diffs, 2 hoodi with allowed_diffs, 2 mainnet clean governance/voting)
- [ ] Verify suggestion output: run any config without its `allowed_diffs` section and confirm the printed snippet can be pasted back to make it pass

### CODE REVIEW

File: diffyscan/diffyscan.py [hunk 1]
```diff
@@ -232,83 +312,111 @@ def run_source_diff(
...
+    identical_files = sum(
+        bool(file_result["file_found"] and not file_result["hunks"])
+        for file_result in file_results
+    )
     logger.info(f"Identical files: {identical_files} / {files_found}")
 
-    files_with_diffs = sum(row[2] and row[3] > 0 for row in report)
-
-    logger.report_table(report)
+    files_with_diffs = sum(bool(file_result["hunks"]) for file_result in file_results)
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply diffyscan/diffyscan.py:404]: haha, I disagree with Copilot here

**Comment by TheDZhon** [comment]
Self-reminder: need to rebase on the latest main and update skills :warning:

File: tests/test_no_wildcard_regression.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        if not path.endswith((".json", ".yaml", ".yml")):
+            continue
+        data = _load(path)
+        if not isinstance(data, dict):
+            continue
+        allowed_diffs = data.get("allowed_diffs") or {}
+        rel = os.path.relpath(path, os.path.join(REPO_ROOT, "config_samples"))
+        for diff_kind in ("source", "bytecode"):
+            for address, rules in (allowed_diffs.get(diff_kind) or {}).items():
+                if not isinstance(rules, list):
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply tests/test_no_wildcard_regression.py]: Fixed in https://github.com/lidofinance/diffyscan/pull/141/commits/a1097db6c673a8b229d7d119593a32e3ee93d372

File: diffyscan/diffyscan.py [hunk 3]
```diff
@@ -170,37 +188,102 @@ def run_bytecode_diff(
...
+    return {
+        "contract_address": contract_address_from_config,
+        "contract_name": contract_name_from_config,
+        "status": evaluation["status"],
+        "match": evaluation["status"] != "failed",
+        "has_diff": True,
+        "matched_rule": evaluation["matched_rule"],
+        "matched_facets": evaluation["matched_facets"],
+        "suggestion_entry": suggestion_entry,
+    }
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply diffyscan/diffyscan.py]: Fixed in https://github.com/lidofinance/diffyscan/pull/141/commits/a1097db6c673a8b229d7d119593a32e3ee93d372

File: diffyscan/utils/node_handler.py [hunk 4]
```diff
@@ -53,22 +53,25 @@ def simulate_deployment(
...
+                    "from": caller,
+                    "to": None,
+                    "gas": hex(gas_limit),
+                    "data": data,
+                },
+                "latest",
+            ],
+        )
+    except NodeError as exc:
+        raise DeploymentSimulationError(str(exc)) from exc
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply diffyscan/utils/node_handler.py]: Fixed in https://github.com/lidofinance/diffyscan/pull/141/commits/a1097db6c673a8b229d7d119593a32e3ee93d372

File: diffyscan/utils/node_handler.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def get_bytecode_from_node(contract_address: str, rpc_url: str) -> str:
     if result == "0x":
         raise NodeError(f"Empty bytecode for contract {contract_address}")
     logger.okay("Bytecode received")
-    return result
+    return str(result)
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply diffyscan/utils/node_handler.py]: Fixed in https://github.com/lidofinance/diffyscan/pull/141/commits/a1097db6c673a8b229d7d119593a32e3ee93d372

**Review by tamtamchik — APPROVED** [review]
👏  Bravo!
